effectful-core/src/Effectful/Exception.hs

```diff
@@ -113,18 +138,436 @@ module Effectful.Exception
+  -> Eff es c
+bracket_ before after action = reallyUnsafeUnliftIO $ \unlift -> do
+  E.bracket_ (unlift before) (unlift after) (unlift action)
+
+-- | Lifted 'E.bracketOnError'.
+bracketOnError
+  :: Eff es a
+  -- ^ Computation to run first.
+  -> (a -> Eff es b)
+  -- ^ Computation to run last when an exception or
+  -- t'Effectful.Error.Static.Error' was thrown.
+  -> (a -> Eff es c)
+  -- ^ Computation to run in-between.
+  -> Eff es c
+bracketOnError before after action = reallyUnsafeUnliftIO $ \unlift -> do
+  E.bracketOnError (unlift before) (unlift . after) (unlift . action)
+
+-- | Generalization of 'bracket'.
+--
+-- See 'C.generalBracket' for more information.
+generalBracket
+  :: Eff es a
+  -- ^ Computation to run first.
+  -> (a -> C.ExitCase c -> Eff es b)
+  -- ^ Computation to run last.
+  -> (a -> Eff es c)
+  -- ^ Computation to run in-between.
+  -> Eff es (c, b)
+generalBracket = C.generalBracket
+
+-- | Lifted 'E.finally'.
+finally
+  :: Eff es a
+  -> Eff es b
+  -- ^ Computation to run last.
+  -> Eff es a
+finally action handler = reallyUnsafeUnliftIO $ \unlift -> do
+  E.finally (unlift action) (unlift handler)
+
+-- | Lifted 'E.onException'.
+onException
+  :: Eff es a
+  -> Eff es b
+  -- ^ Computation to run last when an exception or
+  -- t'Effectful.Error.Static.Error' was thrown.
+  -> Eff es a
+onException action handler = reallyUnsafeUnliftIO $ \unlift -> do
+  E.onException (unlift action) (unlift handler)
+
+----------------------------------------
+-- Utils
+
+-- | Lifted 'E.evaluate'.
 evaluate :: a -> Eff es a
 evaluate = unsafeEff_ . E.evaluate
 
 -- | Deeply evaluate a value using 'evaluate' and 'NFData'.
 evaluateDeep :: NFData a => a -> Eff es a
 evaluateDeep = unsafeEff_ . E.evaluate . force
+
+#if MIN_VERSION_base(4,20,0)
+-- | Lifted 'E.annotateIO'.
+annotateIO :: E.ExceptionAnnotation e => e -> Eff es a -> Eff es a
+annotateIO e action = reallyUnsafeUnliftIO $ \unlift -> do
+  E.annotateIO e (unlift action)
+#endif
+
+----------------------------------------
+-- Check exception type
+
+-- $syncVsAsync
+--
+-- /Note:/ there's no way to determine whether an exception was thrown
+-- synchronously or asynchronously, so these functions rely on a
+-- heuristic. Namely, an exception type is determined by its 'E.Exception'
+-- instance.
+--
+-- >>> import Control.Exception qualified as E
+--
+-- >>> data SyncEx = SyncEx deriving (Show)
+-- >>> instance E.Exception SyncEx
+--
+-- >>> isSyncException SyncEx
+-- Just SyncEx
+--
+-- >>> isAsyncException SyncEx
+-- Nothing
+--
+-- >>> data AsyncEx = AsyncEx deriving (Show)
+-- >>> :{
+--   instance E.Exception AsyncEx where
+--     toException = E.asyncExceptionToException
+--     fromException = E.asyncExceptionFromException
+-- :}
+--
+-- >>> isSyncException AsyncEx
+-- Nothing
+--
+-- >>> isAsyncException AsyncEx
+-- Just AsyncEx
+
+-- | Return 'Just' if the given exception is considered synchronous.
+isSyncException :: E.Exception e => e -> Maybe e
```

Comment:
Hm, shouldn't this be called `maybeSyncException`? I usually expect functions prefixed with "is" to be a predicate; I.e. to return a `Bool`.

Reply:
I agree, though I'd go with `toSyncException`

Reply:
That's even better :+1:

Reply:
I went with option 3 - made `isSyncException` and `isAsyncException` the same type as in `safe-exceptions` and used `catchIf`/`tryIf` for `catchSync`/`trySync` ;) I missed these helpers earlier, that's why I adjusted these functions to fit the `*Just` variant, but that's unnecessary now.